internal/util/cryptsetup.go

```diff
@@ -66,6 +69,134 @@ func LuksStatus(mapperFile string) (string, string, error) {
+		"--key-slot="+slot,
+		"luksAddKey",
+		devicePath,
+		newPassFile.Name(),
+	)
+
+	// Return early if no error to save us some time
+	if err == nil {
+		return nil
+	}
+
+	// Possible scenarios
+	// 1. The provided passphrase to unlock the disk is wrong
+	// 2. The key slot is already in use
+	// 	  If so, check if the key we want to add to the slot is already there
+	//    If not, remove it and then add the new key to the slot
+	if strings.Contains(stderr, fmt.Sprintf("Key slot %s is full", slot)) {
+		// The given slot already has a key
+		// Check if it is the one that we want to update with
+		exists, fErr := LuksVerifyKey(devicePath, newPassphrase, slot)
+		if fErr != nil {
+			return fErr
+		}
+
+		// Verification passed, return early
+		if exists {
+			return nil
+		}
+
+		// Else, we remove the key from the given slot and add the new one
+		// Note: we use existing passphrase here as we are not yet sure if
+		// the newPassphrase is present in the headers
+		fErr = LuksRemoveKey(devicePath, passphrase, slot)
+		if fErr != nil {
+			return fErr
+		}
+
+		// Now the slot is free, add the new key to it
+		fErr = LuksAddKey(devicePath, passphrase, newPassphrase, slot)
+		if fErr != nil {
+			return fErr
+		}
+
+		// No errors, we good.
+		return nil
+	}
+
+	// The existing passphrase is wrong and the slot is empty
+	return err
+}
+
+// LuksRemoveKey removes the key by killing the specified slot.
+func LuksRemoveKey(devicePath, passphrase, slot string) error {
+	keyFile, err := file.CreateTempFile("luks-", passphrase)
+	if err != nil {
+		return err
+	}
+	defer os.Remove(keyFile.Name())
+
+	_, stderr, err := execCryptsetupCommand(
+		nil,
+		"--verbose",
+		"--key-file="+keyFile.Name(),
+		"luksKillSlot",
+		devicePath,
+		slot,
+	)
+	if err != nil {
+		// If a slot is not active, don't treat that as an error
+		if !strings.Contains(stderr, fmt.Sprintf("Keyslot %s is not active.", slot)) {
+			return fmt.Errorf("failed to kill slot %s for device %s: %w", slot, devicePath, err)
+		}
+	}
+
+	return nil
+}
+
+// LuksVerifyKey verifies that a key exists in a given slot.
+func LuksVerifyKey(devicePath, passphrase, slot string) (bool, error) {
+	// Create a temp file that we will use to open the device
+	keyFile, err := file.CreateTempFile("luks-", passphrase)
+	if err != nil {
+		return false, err
+	}
+	defer os.Remove(keyFile.Name())
+
+	// TODO: Should we validate STDOUT?
+	_, stderr, err := execCryptsetupCommand(
+		nil,
+		"--verbose",
+		"--key-file="+keyFile.Name(),
+		"--key-slot="+slot,
+		"luksChangeKey",
+		devicePath,
+		keyFile.Name(),
+	)
+	if err != nil {
+		// If the passphrase doesn't match the key in given slot
+		if strings.Contains(stderr, "No key available with this passphrase.") {
+			// No match, no error
+			return false, nil
+		}
+
+		// Otherwise it was something else, return the wrapped error
+		return false, fmt.Errorf("failed to verify key in slot %s for device %s: %w", slot, devicePath, err)
```

Comment:
Since this is the only instance where we need to log, is it worth it to pass the logger instance to this function as rest of the code in this file is decoupled. If we need to log it anyways, let's include it in the returned error itself?

Reply:
It might help to have the log in case something went wrong. I do not know if the `err` includes (some parts of) `stderr` already. Including the whole `stderr` in `err` might be a bit much, it could be multiple lines.

You can pass the `context.Context` from the caller, and use `internal/util/log.ErrorLog()` or similar, or `ErrorLogMsg()` if you do not want/can pass the context around.

Reply:
> I do not know if the `err` includes (some parts of) `stderr` already.

It does not. I will add a log statement here :)